### PR TITLE
types: Define empty hashes as named constants 

### PR DESCRIPTION
**Pull request by hyunsooda**
## Proposed changes

- Remove scattered literal empty strings
- Defined empty root hash types in `types` package

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

_Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/kaiachain/kaia/blob/main/CONTRIBUTING.md) doc
- [ ] I have read the [CLA](https://gist.github.com/kaiachain-dev/bbf65cc330275c057463c4c94ce787a6) and signed by comment `I have read the CLA Document and I hereby sign the CLA` in first time contribute
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- https://github.com/kaiachain/kaia/issues/142

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by ian0371 — APPROVED** [review]

File: blockchain/types/derive_sha.go [hunk 1]
```diff
@@ -40,9 +41,20 @@ const (
...
+	EmptyRootHashOriginal = commonEmptyRootHash
+
+	// EmptyTxsHash is the known hash of the empty transaction set.
+	EmptyTxsHash = commonEmptyRootHash
+
+	// EmptyReceiptsHash is the known hash of the empty receipt set.
+	EmptyReceiptsHash = commonEmptyRootHash
+
+	// EmptyCodeHash is the known hash of the empty EVM bytecode.
+	EmptyCodeHash = crypto.Keccak256Hash(nil) // c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470
```
Comment (blukat29) [inline blockchain/types/derive_sha.go:57]: Can we have these names instead?

```
// Root hash of an empty trie.
// Used to represent a block's empty StateRoot or an account's StorageRoot.
EmptyRootHash = 0x56e8...

// Hash of an empty byte array.
// Used to represent an account's empty contract code.
EmptyCodeHash = 0xc5d2..

// Empty TransactionRoot and ReceiptsRoot for the Original algorithm.
EmptyTxRootOriginal = EmptyRootHash

// Empty TransactionRoot and ReceiptsRoot for the Simple algorithm.
EmptyTxRootSimple = EmptyCodeHash

// Empty TransactionRoot and ReceiptsRoot for the Simple algorithm.
EmptyTxRootConcat = EmptyCodeHash
```

Comment (hyunsooda) [reply blockchain/types/derive_sha.go:57]: Nice suggestion. The name `EmptyRootHash` has already been declared in L63.
We have two options; changing the function name of L63 or changing the suggested `EmptyRootHash` to another. Which one do you prefer?

File: api/api_ethereum_test.go [hunk 2]
```diff
@@ -230,9 +230,9 @@ func testGetHeader(t *testing.T, testAPIName string, config *params.ChainConfig)
 	header := types.CopyHeader(&types.Header{
 		ParentHash:  common.HexToHash("0xc8036293065bacdfce87debec0094a71dbbe40345b078d21dcc47adb4513f348"),
 		Rewardbase:  common.Address{},
-		TxHash:      common.HexToHash("0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"),
+		TxHash:      types.EmptyCodeHash,
```
Comment (blukat29) [inline api/api_ethereum_test.go:233]:
```suggestion
		TxHash:      types.EmptyTxRootOriginal,
```
Same for other occurences in this file.

File: consensus/gxhash/algorithm_test.go [hunk 3]
```diff
@@ -702,8 +702,8 @@ func TestConcurrentDiskCacheGeneration(t *testing.T) {
 		ParentHash:  common.HexToHash("0xd783efa4d392943503f28438ad5830b2d5964696ffc285f338585e9fe0a37a05"),
 		Coinbase:    common.HexToAddress("0xc0ea08a2d404d3172d2add29a45be56da40e2949"),
 		Root:        common.HexToHash("0x77d14e10470b5850332524f8cd6f69ad21f070ce92dca33ab2858300242ef2f1"),
-		TxHash:      common.HexToHash("0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421"),
-		ReceiptHash: common.HexToHash("0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421"),
+		TxHash:      common.EmptyTxsHash,
+		ReceiptHash: common.EmptyReceiptsHash,
```
Comment (blukat29) [inline consensus/gxhash/algorithm_test.go:706]:
```suggestion
		TxHash:      types. EmptyTxRootOriginal,
		ReceiptHash: types. EmptyTxRootOriginal,
```
Well this code is dead anyway..

File: storage/statedb/iterator.go [hunk 4]
```diff
@@ -149,7 +150,7 @@ func (e seekError) Error() string {
 }
 
 func newNodeIterator(trie *Trie, start []byte) NodeIterator {
-	if trie.Hash() == emptyState {
+	if trie.Hash() == types.EmptyCodeHash {
```
Comment (blukat29) [inline storage/statedb/iterator.go:153]: Can you create `emptyTrieEntry` here in `iterator.go` just for here?

- Geth has EmptyRootHash at this location (https://github.com/ethereum/go-ethereum/blob/c7a8bcecbed7b371f0d6ec97ab1a1ae33aa3ca85/trie/iterator.go#L165). But I don't want to change the existing logic
- `trie.Hash()` is definitely not a code hash
- Therefore we need another name for `0xc5d2..`. Previously it said "emptyState is the known hash of an empty state trie entry.", so emptyTrieEntry is a descriptive name.

File: api/api_ethereum_test.go (hunk 2 above)
Comment (hyeonLewis) [reply api/api_ethereum_test.go:233]: It seems `types.EmptyTxRootSimple` or `types.EmptyTxRootConcat` for `0xc5d2...a470`

File: api/api_ethereum_test.go [hunk 5]
```diff
@@ -230,9 +230,9 @@ func testGetHeader(t *testing.T, testAPIName string, config *params.ChainConfig)
...
 		Root:        common.HexToHash("0xad31c32942fa033166e4ef588ab973dbe26657c594de4ba98192108becf0fec9"),
-		ReceiptHash: common.HexToHash("0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"),
+		ReceiptHash: types.EmptyCodeHash,
```
Comment (hyeonLewis) [inline api/api_ethereum_test.go:235]:
```suggestion
		ReceiptHash: types.EmptyTxRootSimple,
```
Also please reflect this for other `ReceiptHash`.

File: blockchain/types/derive_sha.go [hunk 6]
```diff
@@ -40,15 +41,32 @@ const (
...
+
+	// Empty TransactionRoot and ReceiptsRoot for the Original algorithm.
+	EmptyTxRootOriginal = EmptyRootHash
+
+	// Empty TransactionRoot and ReceiptsRoot for the Simple algorithm.
+	EmptyTxRootSimple = EmptyCodeHash
+
+	// Empty TransactionRoot and ReceiptsRoot for the Concat algorithm.
+	EmptyTxRootConcat = EmptyCodeHash
```
Comment (hyeonLewis) [inline blockchain/types/derive_sha.go:61]: How about using `EmptyDeriveShaOrig`, `EmptyDeriveShaSimple`, and `EmptyDeriveShaConcat` for clarity?

File: api/api_ethereum_test.go (hunk 2 above)
Comment (hyeonLewis) [reply api/api_ethereum_test.go:233]: It seems `types.EmptyTxRootSimple` or `types.EmptyTxRootConcat` for `0xc5d2...a470`.

Comment (hyunsooda) [reply api/api_ethereum_test.go:233]: @blukat29 @hyeonLewis I think that `TxHash` can be modified to `types.EmptyTxRootOriginal` , which is the expected value. The result of different hash should be recalculated accordingly.

File: blockchain/types/derive_sha.go (hunk 6 above)
Comment (blukat29) [reply blockchain/types/derive_sha.go:61]: I'd prefer TxRoot. I did think about TxReceiptRoot or TxAndReceiptRoot but both are wordy. On the other hand, DeriveSha sounds too generic to me.

File: api/api_ethereum_test.go (hunk 2 above)
Comment (blukat29) [reply api/api_ethereum_test.go]: If it's not too important (judging from the comment "dummy header") then I think EmptyTxRootOriginal is more representative "dummy" example.

But if the test ever fails, then please go back to the ones (Simple or Concat) equivalent to the original value (0xc5d2..).

Comment (hyunsooda) [reply api/api_ethereum_test.go]: PTAL https://github.com/kaiachain/kaia/pull/146/commits/00cd23eafa9159337e3fe6266d3f901c087c19f8

File: blockchain/types/derive_sha.go (hunk 6 above)
Comment (hyeonLewis) [reply blockchain/types/derive_sha.go:61]: Ok, then let's keep the names. Thanks.

File: api/api_ethereum_test.go [hunk 7]
```diff
@@ -334,9 +334,9 @@ func testGetBlock(t *testing.T, testAPIName string, fullTxs bool) {
 
 	// Create dummy header
 	header := types.CopyHeader(&types.Header{
-		ParentHash: common.HexToHash("0xc8036293065bacdfce87debec0094a71dbbe40345b078d21dcc47adb4513f348"), Rewardbase: common.Address{}, TxHash: common.HexToHash("0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"),
+		ParentHash: common.HexToHash("0xc8036293065bacdfce87debec0094a71dbbe40345b078d21dcc47adb4513f348"), Rewardbase: common.Address{}, TxHash: types.EmptyCodeHash,
```
Comment (blukat29) [inline api/api_ethereum_test.go:337]: TxHash is not a kind of CodeHash.
```suggestion
		ParentHash: common.HexToHash("0xc8036293065bacdfce87debec0094a71dbbe40345b078d21dcc47adb4513f348"), Rewardbase: common.Address{}, TxHash: types.EmptyTxRootOriginal,
```

File: api/api_ethereum_test.go [hunk 8]
```diff
@@ -334,9 +334,9 @@ func testGetBlock(t *testing.T, testAPIName string, fullTxs bool) {
...
 		Root:        common.HexToHash("0xad31c32942fa033166e4ef588ab973dbe26657c594de4ba98192108becf0fec9"),
-		ReceiptHash: common.HexToHash("0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"),
+		ReceiptHash: types.EmptyCodeHash,
```
Comment (blukat29) [inline api/api_ethereum_test.go:339]:
```suggestion
		ReceiptHash: types.EmptyTxRootOriginal,
```

File: api/api_ethereum_test.go [hunk 9]
```diff
@@ -230,9 +230,9 @@ func testGetHeader(t *testing.T, testAPIName string, config *params.ChainConfig)
 	header := types.CopyHeader(&types.Header{
 		ParentHash:  common.HexToHash("0xc8036293065bacdfce87debec0094a71dbbe40345b078d21dcc47adb4513f348"),
 		Rewardbase:  common.Address{},
-		TxHash:      common.HexToHash("0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"),
+		TxHash:      types.EmptyTxRootOriginal,
 		Root:        common.HexToHash("0xad31c32942fa033166e4ef588ab973dbe26657c594de4ba98192108becf0fec9"),
-		ReceiptHash: common.HexToHash("0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"),
+		ReceiptHash: types.EmptyTxRootSimple,
```
Comment (blukat29) [inline api/api_ethereum_test.go:235]:
```suggestion
		ReceiptHash: types.EmptyTxRootOriginal,
```
While it doesn't matter in the test case, it is unrealistic to have two different DeriveSha algorithms are used in one header.

**Comment by hyunsooda** [comment]
@blukat29 All set. The hash values look correctly assigned for each specific purpose.

**Review by blukat29 — APPROVED** [review]

**Review by hyeonLewis — APPROVED** [review]

**Pull request merged**